File: packages/domain/src/time_block.ts
import { DateTime } from 'luxon';
import { Item, type ItemDto } from './item';

export interface TimeBlockDto {
    block_id:  number;
    item:      ItemDto;
    date:      string;
    start_min: number;
    end_min:   number;
    note:      string;
}

export interface CreateTimeBlockDto {
    item_id:   number;
    date:      string;
    start_min: number;
    end_min:   number;
    note?:     string;
}

export interface UpdateTimeBlockDto {
    block_id:   number;
    date?:      string;
    start_min?: number;
    end_min?:   number;
    note?:      string;
}

function minToLabel(min: number): string {
    const h = Math.floor(min / 60).toString().padStart(2, '0');
    const m = (min % 60).toString().padStart(2, '0');
    return `${h}:${m}`;
}

export class TimeBlock {
    public BlockId:  number;
    public Item:     Item;
    public Date:     DateTime;
    public StartMin: number;
    public EndMin:   number;
    public Note:     string;

    public constructor(dto: TimeBlockDto) {
        this.BlockId  = dto.block_id;
        this.Item     = new Item(dto.item);
        this.Date     = DateTime.fromISO(dto.date);
        this.StartMin = dto.start_min;
        this.EndMin   = dto.end_min;
        this.Note     = dto.note;
    }

    public startLabel(): string {
        return minToLabel(this.StartMin);
    }

    public endLabel(): string {
        return minToLabel(this.EndMin);
    }
}

File: crates/zealot-domain/src/time_block.rs
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};

use crate::{common::id::Id, item::{Item, ItemDto}};

pub struct TimeBlock {
    pub block_id:  Id,
    pub item:      Item,
    pub date:      NaiveDate,
    pub start_min: i32,
    pub end_min:   i32,
    pub note:      String,
}

/// Lightweight time block returned directly by the repo.
/// The service layer hydrates the full Item.
#[derive(Debug, Clone)]
pub struct TimeBlockCore {
    pub block_id:  Id,
    pub item_id:   Id,
    pub date:      NaiveDate,
    pub start_min: i32,
    pub end_min:   i32,
    pub note:      String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TimeBlockDto {
    pub block_id:  i64,
    pub item:      ItemDto,
    pub date:      String,
    pub start_min: i32,
    pub end_min:   i32,
    pub note:      String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct CreateTimeBlockDto {
    pub item_id:   i64,
    pub date:      String,
    pub start_min: i32,
    pub end_min:   i32,
    pub note:      Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct UpdateTimeBlockDto {
    pub block_id:  i64,
    pub date:      Option<String>,
    pub start_min: Option<i32>,
    pub end_min:   Option<i32>,
    pub note:      Option<String>,
}

impl From<TimeBlock> for TimeBlockDto {
    fn from(b: TimeBlock) -> Self {
        Self {
            block_id:  b.block_id.into(),
            item:      ItemDto::from(&b.item),
            date:      b.date.format("%Y-%m-%d").to_string(),
            start_min: b.start_min,
            end_min:   b.end_min,
            note:      b.note,
        }
    }
}
